### PR TITLE
Name resolution 2.0

### PR DESCRIPTION
**Pull request by dima74**
Name resolution 2.0 implementaion. Please see [meta issue](https://github.com/intellij-rust/intellij-rust/issues/6217) for details.

The new name resolution is disabled by default.

Please, note that the new resolve engine is in an experimental state and has known regressions. It is intended for internal use. If you wish to give a try:
1. Ensure you have at lease `-Xmx1500m` in `Help | Edit custom VM Options`
2. Enable `org.rust.resolve.new.engine` registry key
3. Enable `org.rust.lang.macros.persistentCache` registry key

### CODE REVIEW

File: src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt [hunk 1]
```diff
@@ -892,31 +896,55 @@ private class MacroResolver private constructor(private val processor: RsResolve
...
-        val context = expandedFrom ?: element.context ?: return false
+        if (expandedFrom != null && processExpandedFrom(expandedFrom)) return MacroResolveResult.True
+        val context = expandedFrom ?: element.context ?: return MacroResolveResult.False
+
+        if (context is RsMod) {
+            /** [processRemainedExportedMacros] processes local imports */
+            val result = processMacros(context, processor, ::processRemainedExportedMacros)
+            if (result !== null) return result.toResult(usedNewResolve = true)
+        }
+        for (e in element.leftSiblings) {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt:907]: This reordering of `processExpandedFrom` and `for (e in element.leftSiblings)` looks breaking in a case like this
```
foo! {
    macro_rules! foo { ... }
    foo!();
}
```

File: src/main/kotlin/org/rust/lang/core/resolve2/FileModificationTracker.kt [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            crate,
+            hashCalculator,
+            fileRelativePath,
+            collectChildModules = true
+        )
+        ModCollectorBase.collectMod(mod, isDeeplyEnabledByCfg, visitor, crate)
+    }
+}
+
+private fun <T : Comparable<T>> compareArrays(a: Array<T>, b: Array<T>): Int {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve2/FileModificationTracker.kt:155]: `Arrays.compare` ?

File: src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+        val perNs = PerNs(visItem, item.namespaces)
+        onAddItem(modData, name, perNs)
+
+        // we have to check `modData[name]` to be sure that `childModules` and `visibleItems` are consistent
+        if (childModData !== null && perNs.types === modData[name].types) {
+            modData.childModules[name] = childModData
+        }
+    }
+
+    private fun convertToVisItem(item: ItemLight, stub: RsNamedStub): VisItem? {
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt:121]:
```suggestion
    private fun convertToVisItem(item: ItemLight, stub: RsNamedStub): VisItem {
```

File: src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt [hunk 4]
```diff
@@ -0,0 +1,362 @@
...
+            ownedDirectoryId = modData.ownedDirectoryId,  // actually can use any value here
+            isEnum = true
+        )
+        for (variantPsi in enumStub.variants) {
+            val variantName = variantPsi.name ?: continue
+            val variantPath = enumPath.append(variantName)
+            val isVariantDeeplyEnabledByCfg = enumData.isDeeplyEnabledByCfg && variantPsi.isEnabledByCfgSelf(crate)
+            val variantVisibility = if (isVariantDeeplyEnabledByCfg) Visibility.Public else Visibility.CfgDisabled
+            val variant = VisItem(variantPath, variantVisibility)
+            enumData.visibleItems[variantName] = PerNs(variant, variantPsi.namespaces)
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt:199]: Can cfg-disabled variant override cfg-enabled here?

File: src/main/kotlin/org/rust/lang/core/resolve2/DefMapMultithreadBuilder.kt [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.util.Computable
+import org.rust.lang.core.crate.Crate
+import org.rust.openapiext.executeUnderProgress
+import java.util.concurrent.*
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.system.measureTimeMillis
+
+/** Builds [CrateDefMap] for [crates] in parallel using [pool] and with respect to dependency graph */
+class DefMapMultithreadBuilder(
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/DefMapMultithreadBuilder.kt]: Renamed to `DefMapsBuilder`

File: src/main/kotlin/org/rust/lang/core/resolve2/DefMapMultithreadBuilder.kt [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            }
+            .toMap(hashMapOf())
+        val defMap = buildDefMap(crate, allDependenciesDefMaps)
+        val holder = defMapService.getDefMapHolder(crateId)
+        holder.defMap = defMap
+        holder.shouldRebuild = false
+        holder.setLatestStamp()
+        if (defMap !== null) {
+            builtDefMaps[crate] = defMap
+        }
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/DefMapMultithreadBuilder.kt]: Added method `DefMapService#setDefMap`

File: src/main/kotlin/org/rust/lang/core/resolve2/DefMapService.kt [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+import org.rust.openapiext.checkWriteAccessAllowed
+import org.rust.openapiext.pathAsPath
+import org.rust.stdext.mapToSet
+import java.nio.file.Path
+import java.util.concurrent.ConcurrentHashMap
+
+class DefMapHolder(private val structureModificationTracker: ModificationTracker) {
+
+    @Volatile
+    var defMap: CrateDefMap? = null
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/DefMapService.kt:45]: Added some comments. Also setters for fields `defMap` and `defMapStamp` in `DefMapHolder` are private now

File: src/main/kotlin/org/rust/lang/core/resolve2/DefMapService.kt [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    var shouldRecheck: Boolean = false
+    val changedFiles: MutableSet<RsFile> = hashSetOf()
+
+    override fun toString(): String = "DefMapHolder($defMap, stamp=$defMapStamp)"
+}
+
+@Service
+class DefMapService(val project: Project) : Disposable {
+
+    private val defMaps: ConcurrentHashMap<CratePersistentId, DefMapHolder> = ConcurrentHashMap()
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/DefMapService.kt:129]: `DefMapsBuilder` uses multiple threads for building defMaps, so I think `ConcurrentHashMap` is still needed

File: src/main/kotlin/org/rust/lang/core/resolve2/FacadeBuildDefMap.kt [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+fun buildDefMap(crate: Crate, allDependenciesDefMaps: Map<Crate, CrateDefMap>): CrateDefMap? {
+    // TODO: What if dumb mode?
+    checkReadAccessAllowed()
+    val project = crate.project
+    check(isNewResolveEnabled)
+    val context = CollectorContext(crate, project)
+    val defMap = buildDefMapContainingExplicitItems(context, allDependenciesDefMaps) ?: return null
+    DefCollector(project, defMap, context).collect()
+    defMap.afterBuilt()
+    project.defMapService.afterDefMapBuilt(defMap)
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/FacadeBuildDefMap.kt]: Moved to `DefMapService#setDefMap`

File: src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+     * called when new [RsItemElement] is found
+     * default behaviour: just add it to [ModData.visibleItems]
+     * behaviour when processing expanded items:
+     * add it to [ModData.visibleItems] and propagate to modules which have glob import from [ModData]
+     */
+    private val onAddItem: (ModData, String, PerNs) -> Unit =
+        { containingMod, name, perNs -> containingMod.addVisibleItem(name, perNs) },
+) : ModVisitor {
+
+    private var hashCalculator: HashCalculator? = null
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt]: Added `ModCollectorContext` and two public methods `collectFileAndCalculateHash` and `collectExpandedElements`. And `ModCollector` is now private and receives `hashCalculator` in constructor

File: src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+    fun collectFileAndCalculateHash(file: RsFile) {
+        val hashCalculator = HashCalculator()
+        collectMod(file.getStubOrBuild() ?: return, hashCalculator)
+        val fileHash = hashCalculator.getFileHash()
+        defMap.addVisitedFile(file, modData, fileHash)
+    }
+
+    fun collectExpandedElements(expandedFile: RsFileStub) = collectMod(expandedFile, null)
+
+    private fun collectMod(mod: StubElement<out RsMod>, hashCalculator: HashCalculator?) {
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt]: We have only two occurrences of `StubElement<out RsMod>`, so I am not sure that it is worth it

File: src/main/kotlin/org/rust/lang/core/resolve2/FacadeResolve.kt [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+                ?: return null
+        }
+}
+
+// TODO: Multiresolve
+private inline fun <reified T : RsElement> List<T>.singleOrCfgEnabled(): T? =
+    singleOrNull() ?: singleOrNull { it.isEnabledByCfg }
+
+/** [expandedItemsExceptImplsAndUses] with addition of items from [RsForeignModItem]s */
+private val RsItemsOwner.expandedItemsAll: List<RsItemElement>
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/FacadeResolve.kt:369]: I think there rarely will be any `RsForeignModItem`, so current code is optimized for it

File: src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt (hunk 4 above)
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/ModCollector.kt]: Thanks, fixed

File: src/main/kotlin/org/rust/lang/core/resolve2/FacadeBuildDefMap.kt [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+import org.rust.openapiext.checkReadAccessAllowed
+import org.rust.openapiext.fileId
+import org.rust.openapiext.testAssert
+
+/**
+ * Returns `null` if [crate] has null `id` or `rootMod`,
+ * or if crate should not be indexed (e.g. test/bench non-workspace crate)
+ */
+fun buildDefMap(crate: Crate, allDependenciesDefMaps: Map<Crate, CrateDefMap>): CrateDefMap? {
+    // TODO: What if dumb mode?
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve2/FacadeBuildDefMap.kt:26]: The comment is still here :)

File: src/main/kotlin/org/rust/lang/core/resolve2/FacadeResolve.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve2/FacadeResolve.kt:369]: But its clients also use stream-like operations, e.g. this code produces 2 intermediate collections:
```
mod.expandedItemsAll
                .filterIsInstance<RsModItem>()
                .filter { it.modName == segment }
                .singleOrCfgEnabled()
```

File: src/main/kotlin/org/rust/lang/core/resolve2/FacadeBuildDefMap.kt [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+private fun findPrelude(crate: Crate, allDependenciesDefMaps: Map<Crate, CrateDefMap>): ModData? {
+    /**
+     * Correct prelude is always selected (core vs std), because [Crate.flatDependencies] is top sorted.
+     * And we peek prelude from latest dependency.
+     */
+    val dependencies = crate.dependencies.mapToSet { it.crate }
+    val dependenciesTopSorted = crate.flatDependencies.filter { it in dependencies }
+    return dependenciesTopSorted
+        .mapNotNull { allDependenciesDefMaps[it]?.prelude }
+        .lastOrNull()
```
[the remark this answers is not in the document]
Comment (dima74) [reply src/main/kotlin/org/rust/lang/core/resolve2/FacadeBuildDefMap.kt]: Replaced to `.asReversed().asSequence().mapNotNull { ... }.firstOrNull()` - unfortunatelly there is not method `asReversed` for `LinkedHashSet`

**Comment by matklad** [comment]
:fire: :fire: :fire:  kudos @dima74 !

**Pull request merged**